actors/dagger.py

```diff
@@ -0,0 +1,356 @@
+            observations = []
+            actions = []
+
+            # Rollout a bunch of trials to gather expert data
+            for (step_idx, _trial_id, _trial_idx, sample,) in run_session.start_and_await_trials(
+                trials_id_and_params=[
+                    (f"{run_session.run_id}_{trial_idx}", create_trial_params_gather_expert_data(trial_idx))
+                    for trial_idx in range(self._cfg.num_data_gather_trials)
+                ],
+                sample_producer_impl=self.sample_producer,
+                num_parallel_trials=self._cfg.num_parallel_trials,
+            ):
+                (teacher_observation, teacher_action, _) = sample
+
+                if (_trial_idx + 1) % 10 == 0:
+                    log.info(f"Gathering expert data, trial {_trial_idx + 1}/{self._cfg.num_data_gather_trials}")
+
+                actions.extend(teacher_action)
+                observations.extend(teacher_observation)
+
+            # Step 2: Teach the student algorithm using DAGGER
+            optimizer = torch.optim.Adam(
+                student_model.policy_network.parameters(),
+                lr=self._cfg.learning_rate,
+            )
+            loss_fn = torch.nn.CrossEntropyLoss()
+
+            # Load the teacher model
+            teacher_model, _, _ = await run_session.model_registry.retrieve_version(
+                SimpleA2CModel, self._cfg.teacher_model_id, -1
+            )
+
+            # Rollout a bunch of trials to train the student model
+            for (step_idx, _trial_id, _trial_idx, sample,) in run_session.start_and_await_trials(
+                trials_id_and_params=[
+                    (f"{run_session.run_id}_{trial_idx}", create_trial_params_imitatation(trial_idx))
+                    for trial_idx in range(self._cfg.num_imitation_trials)
+                ],
+                sample_producer_impl=self.sample_producer,
+                num_parallel_trials=self._cfg.num_parallel_trials,
+            ):
+                (student_observations, _, student_rewards) = sample
+
+                if (_trial_idx + 1) % 10 == 0:
+                    log.info(f"Training the student, trial {_trial_idx + 1}/{self._cfg.num_imitation_trials}")
+
+                for student_observation in student_observations:
+                    # Feed the student's observation to the teacher model to find the correct action
+                    probs = torch.softmax(teacher_model.actor_network(student_observation), dim=-1)
+                    discrete_action_tensor = torch.distributions.Categorical(probs).sample()
+                    action_value = SpaceValue(
+                        properties=[SpaceValue.PropertyValue(discrete=discrete_action_tensor.item())]
+                    )
+                    correct_action = torch.tensor(
+                        flatten(self._environment_specs.action_space, action_value), dtype=self._dtype
+                    )
+                    actions.append(correct_action)
```

Comment:
Would be easier to follow if this one is named "corrected_actions" or "teacher_actions"

Reply:
you might also want to use/adapt https://github.com/cogment/cogment-verse/blob/next/cogment_verse/replay_buffers/torch_replay_buffer.py

Reply:
Added a linear schedule similar to the simple_dqn epsilon. I think I'll skip using the replay buffer since I don't need all that info to be stored.